File: tl/src/main/java/com/github/badoualy/telegram/tl/api/request/TLRequestPaymentsSendPaymentForm.java
package com.github.badoualy.telegram.tl.api.request;

import com.github.badoualy.telegram.tl.TLContext;
import com.github.badoualy.telegram.tl.api.TLAbsInputPaymentCredentials;
import com.github.badoualy.telegram.tl.api.payments.TLAbsPaymentResult;
import com.github.badoualy.telegram.tl.core.TLMethod;
import com.github.badoualy.telegram.tl.core.TLObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readInt;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLObject;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeInt;
import static com.github.badoualy.telegram.tl.StreamUtils.writeString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLObject;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32;
import static com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize;

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
public class TLRequestPaymentsSendPaymentForm extends TLMethod<TLAbsPaymentResult> {

    public static final int CONSTRUCTOR_ID = 0x2b8879b3;

    protected int flags;

    protected int msgId;

    protected String requestedInfoId;

    protected String shippingOptionId;

    protected TLAbsInputPaymentCredentials credentials;

    private final String _constructor = "payments.sendPaymentForm#2b8879b3";

    public TLRequestPaymentsSendPaymentForm() {
    }

    public TLRequestPaymentsSendPaymentForm(int msgId, String requestedInfoId, String shippingOptionId, TLAbsInputPaymentCredentials credentials) {
        this.msgId = msgId;
        this.requestedInfoId = requestedInfoId;
        this.shippingOptionId = shippingOptionId;
        this.credentials = credentials;
    }

    @Override
    @SuppressWarnings({"unchecked", "SimplifiableConditionalExpression"})
    public TLAbsPaymentResult deserializeResponse(InputStream stream, TLContext context) throws IOException {
        final TLObject response = readTLObject(stream, context);
        if (response == null) {
            throw new IOException("Unable to parse response");
        }
        if (!(response instanceof TLAbsPaymentResult)) {
            throw new IOException(
                    "Incorrect response type, expected " + getClass().getCanonicalName() + ", found " + response
                            .getClass().getCanonicalName());
        }
        return (TLAbsPaymentResult) response;
    }

    private void computeFlags() {
        flags = 0;
        flags = requestedInfoId != null ? (flags | 1) : (flags & ~1);
        flags = shippingOptionId != null ? (flags | 2) : (flags & ~2);
    }

    @Override
    public void serializeBody(OutputStream stream) throws IOException {
        computeFlags();

        writeInt(flags, stream);
        writeInt(msgId, stream);
        if ((flags & 1) != 0) {
            if (requestedInfoId == null) throwNullFieldException("requestedInfoId", flags);
            writeString(requestedInfoId, stream);
        }
        if ((flags & 2) != 0) {
            if (shippingOptionId == null) throwNullFieldException("shippingOptionId", flags);
            writeString(shippingOptionId, stream);
        }
        writeTLObject(credentials, stream);
    }

    @Override
    @SuppressWarnings({"unchecked", "SimplifiableConditionalExpression"})
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {
        flags = readInt(stream);
        msgId = readInt(stream);
        requestedInfoId = (flags & 1) != 0 ? readTLString(stream) : null;
        shippingOptionId = (flags & 2) != 0 ? readTLString(stream) : null;
        credentials = readTLObject(stream, context, TLAbsInputPaymentCredentials.class, -1);
    }

    @Override
    public int computeSerializedSize() {
        computeFlags();

        int size = SIZE_CONSTRUCTOR_ID;
        size += SIZE_INT32;
        size += SIZE_INT32;
        if ((flags & 1) != 0) {
            if (requestedInfoId == null) throwNullFieldException("requestedInfoId", flags);
            size += computeTLStringSerializedSize(requestedInfoId);
        }
        if ((flags & 2) != 0) {
            if (shippingOptionId == null) throwNullFieldException("shippingOptionId", flags);
            size += computeTLStringSerializedSize(shippingOptionId);
        }
        size += credentials.computeSerializedSize();
        return size;
    }

    @Override
    public String toString() {
        return _constructor;
    }

    @Override
    public int getConstructorId() {
        return CONSTRUCTOR_ID;
    }

    public int getMsgId() {
        return msgId;
    }

    public void setMsgId(int msgId) {
        this.msgId = msgId;
    }

    public String getRequestedInfoId() {
        return requestedInfoId;
    }

    public void setRequestedInfoId(String requestedInfoId) {
        this.requestedInfoId = requestedInfoId;
    }

    public String getShippingOptionId() {
        return shippingOptionId;
    }

    public void setShippingOptionId(String shippingOptionId) {
        this.shippingOptionId = shippingOptionId;
    }

    public TLAbsInputPaymentCredentials getCredentials() {
        return credentials;
    }

    public void setCredentials(TLAbsInputPaymentCredentials credentials) {
        this.credentials = credentials;
    }
}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/request/TLRequestPaymentsSendPaymentForm.kt
package com.github.badoualy.telegram.tl.api.request

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_BOOLEAN
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_DOUBLE
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLBytesSerializedSize
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize
import com.github.badoualy.telegram.tl.api.TLAbsInputPaymentCredentials
import com.github.badoualy.telegram.tl.api.TLInputPaymentCredentialsApplePay
import com.github.badoualy.telegram.tl.api.payments.TLAbsPaymentResult
import com.github.badoualy.telegram.tl.core.TLMethod
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException
import kotlin.Any
import kotlin.Boolean
import kotlin.Int
import kotlin.String
import kotlin.jvm.Throws

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLRequestPaymentsSendPaymentForm() : TLMethod<TLAbsPaymentResult>() {
    var msgId: Int = 0

    var requestedInfoId: String? = null

    var shippingOptionId: String? = null

    var credentials: TLAbsInputPaymentCredentials = TLInputPaymentCredentialsApplePay()

    private val _constructor: String = "payments.sendPaymentForm#2b8879b3"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(
            msgId: Int,
            requestedInfoId: String?,
            shippingOptionId: String?,
            credentials: TLAbsInputPaymentCredentials
    ) : this() {
        this.msgId = msgId
        this.requestedInfoId = requestedInfoId
        this.shippingOptionId = shippingOptionId
        this.credentials = credentials
    }

    protected override fun computeFlags() {
        _flags = 0
        updateFlags(requestedInfoId, 1)
        updateFlags(shippingOptionId, 2)
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        computeFlags()

        writeInt(_flags)
        writeInt(msgId)
        doIfMask(requestedInfoId, 1) { writeString(it) }
        doIfMask(shippingOptionId, 2) { writeString(it) }
        writeTLObject(credentials)
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        _flags = readInt()
        msgId = readInt()
        requestedInfoId = readIfMask(1) { readString() }
        shippingOptionId = readIfMask(2) { readString() }
        credentials = readTLObject<TLAbsInputPaymentCredentials>()
    }

    override fun computeSerializedSize(): Int {
        computeFlags()

        var size = SIZE_CONSTRUCTOR_ID
        size += SIZE_INT32
        size += SIZE_INT32
        size += getIntIfMask(requestedInfoId, 1) { computeTLStringSerializedSize(it) }
        size += getIntIfMask(shippingOptionId, 2) { computeTLStringSerializedSize(it) }
        size += credentials.computeSerializedSize()
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLRequestPaymentsSendPaymentForm) return false
        if (other === this) return true

        return _flags == other._flags
                && msgId == other.msgId
                && requestedInfoId == other.requestedInfoId
                && shippingOptionId == other.shippingOptionId
                && credentials == other.credentials
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0x2b8879b3.toInt()
    }
}
